java/java.nativeimage.debugger/src/org/netbeans/modules/java/nativeimage/debugger/displayer/JavaFrameDisplayer.java

```diff
@@ -161,35 +162,44 @@ private static ClassPath findSourcePath(Project project) {
             List<FileObject> allSourceRoots = new ArrayList<>();
             Set<FileObject> preferredRoots = new HashSet<>();
             Set<FileObject> addedBinaryRoots = new HashSet<>();
-            SourceGroup[] sgs = ProjectUtils.getSources(project).getSourceGroups(JavaProjectConstants.SOURCES_TYPE_JAVA);
-            for (SourceGroup sg : sgs) {
-                ClassPath ecp = ClassPath.getClassPath(sg.getRootFolder(), ClassPath.EXECUTE);
-                if (ecp == null) {
-                    ecp = ClassPath.getClassPath(sg.getRootFolder(), ClassPath.SOURCE);
-                }
-                if (ecp != null) {
-                    FileObject[] binaryRoots = ecp.getRoots();
-                    for (FileObject fo : binaryRoots) {
-                        if (addedBinaryRoots.contains(fo)) {
-                            continue;
-                        }
-                        addedBinaryRoots.add(fo);
-                        FileObject[] roots = SourceForBinaryQuery.findSourceRoots(fo.toURL()).getRoots();
-                        for (FileObject fr : roots) {
-                            if (!preferredRoots.contains(fr)) {
-                                allSourceRoots.add(fr);
-                                preferredRoots.add(fr);
-                            }
-                        }
-                    }
-                }
+            Sources sources = ProjectUtils.getSources(project);
+            SourceGroup[] sgs = sources.getSourceGroups(JavaProjectConstants.SOURCES_TYPE_JAVA);
+            addSourceRoots(sgs, allSourceRoots, preferredRoots, addedBinaryRoots);
+            sgs = sources.getSourceGroups(JavaProjectConstants.SOURCES_TYPE_GENERATED);
+            if (sgs != null) {
+                addSourceRoots(sgs, allSourceRoots, preferredRoots, addedBinaryRoots);
             }
             return createClassPath(allSourceRoots);
         } else {
             return null;
         }
     }
 
+    private static void addSourceRoots(SourceGroup[] sgs, List<FileObject> allSourceRoots, Set<FileObject> preferredRoots, Set<FileObject> addedBinaryRoots) {
+        for (SourceGroup sg : sgs) {
+            ClassPath ecp = ClassPath.getClassPath(sg.getRootFolder(), ClassPath.EXECUTE);
+            if (ecp == null) {
+                ecp = ClassPath.getClassPath(sg.getRootFolder(), ClassPath.SOURCE);
+            }
+            if (ecp != null) {
+                FileObject[] binaryRoots = ecp.getRoots();
+                for (FileObject fo : binaryRoots) {
+                    if (addedBinaryRoots.contains(fo)) {
+                        continue;
+                    }
+                    addedBinaryRoots.add(fo);
+                    FileObject[] roots = SourceForBinaryQuery.findSourceRoots(fo.toURL()).getRoots();
+                    for (FileObject fr : roots) {
+                        if (!preferredRoots.contains(fr)) {
```

Comment:
It seems that `preferredSourceRoots` is just an optimization for `contains` -- what if the `allSourceRoots` would be a `LinkedHashMap` ?